contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,527 @@
+  }
+
+  function getRandomnessFromProof(
+    bytes memory proof
+  )
+    public 
+    view 
+    returns (
+      bytes32 currentKeyHash,
+      uint256 requestId, 
+      uint256 randomness, 
+      FulfillmentParams memory fp
+    ) 
+  {
+    // blockNum follows proof, which follows length word (only direct-number
+    // constants are allowed in assembly, so have to compute this in code)
+    uint256 blockNumOffset = 0x20 + PROOF_LENGTH;
+    // Note that proof.length skips the initial length word.
+    // We expected the total length to be proof + 5 words (blocknum, subId, callbackLimit, nw, sender)
+    if (proof.length != PROOF_LENGTH + 0x20*5) {
+      revert InvalidProofLength(proof.length, PROOF_LENGTH + 0x20*5);
+    }
+    uint256[2] memory publicKey;
+    uint256 preSeed;
+    uint256 blockNum;
+    address sender;
+    assembly { // solhint-disable-line no-inline-assembly
+      publicKey := add(proof, PUBLIC_KEY_OFFSET)
+      preSeed := mload(add(proof, PRESEED_OFFSET))
+      blockNum := mload(add(proof, blockNumOffset))
+      // We use a struct to limit local variables to avoid stack depth errors.
+      mstore(fp, mload(add(add(proof, blockNumOffset), 0x20)))
+      mstore(add(fp, 0x20), mload(add(add(proof, blockNumOffset), 0x40)))
+      mstore(add(fp, 0x40), mload(add(add(proof, blockNumOffset), 0x60)))
+      sender := mload(add(add(proof, blockNumOffset), 0x80))
+    }
+    currentKeyHash = hashOfKey(publicKey);
+    bytes32 callback = s_callbacks[preSeed];
+    requestId = preSeed;
+    if (callback == 0) {
+      revert NoCorrespondingRequest();
+    }
+    if (callback != keccak256(abi.encodePacked(requestId, blockNum, fp.subId, fp.callbackGasLimit, fp.numWords, sender))) {
+      revert IncorrectCommitment();
+    }
+    fp.sender = sender;
+
+    bytes32 blockHash = blockhash(blockNum);
+    if (blockHash == bytes32(0)) {
+      blockHash = BLOCKHASH_STORE.getBlockhash(blockNum);
+      if (blockHash == bytes32(0)) {
+        revert BlockHashNotInStore(blockNum);
+      }
+    }
+    // The seed actually used by the VRF machinery, mixing in the blockhash
+    uint256 actualSeed = uint256(keccak256(abi.encodePacked(preSeed, blockHash)));
+    // solhint-disable-next-line no-inline-assembly
+    assembly { // Construct the actual proof from the remains of proof
+      mstore(add(proof, PRESEED_OFFSET), actualSeed)
+      mstore(proof, PROOF_LENGTH)
+    }
+    randomness = VRF.randomValueFromVRFProof(proof); // Reverts on failure
+  }
+
+  function getFeedData()
+    private
+    view
+    returns (
+        uint256
+    )
+  {
+    uint32 stalenessSeconds = s_config.stalenessSeconds;
+    bool staleFallback = stalenessSeconds > 0;
+    uint256 timestamp;
+    int256 linkEth;
+    (,linkEth,,timestamp,) = LINK_ETH_FEED.latestRoundData();
+    if (staleFallback && stalenessSeconds < block.timestamp - timestamp) {
```

Comment:
Why do we use a value set at config time, if the feed price is older than the staleness threshold? Wouldn't an old price usually still be closer to the correct price than one set at config time?

Reply:
This logic was actually ported from the keeper contracts. Its a good question - @se3000 whats the story with the fallback link price?

Reply:
spoke offline